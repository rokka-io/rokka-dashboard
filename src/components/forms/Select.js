import PropTypes from 'prop-types'
import React from 'react'

const Select = ({ value, defaultValue, placeholder, disabled, children, ...props }) => {
  const ph = defaultValue
    ? `${placeholder || props.name || props.type || ''} (default: ${defaultValue})`
    : placeholder

  const val = disabled || !props.onChange
    ? (value || defaultValue)
    : value

  return <select {...props} value={val} placeholder={ph} disabled={disabled || !props.onChange}>
    <option key="__default">{ph}</option>
    {children}
  </select>
}

Select.propTypes = {
  value: PropTypes.any,
  defaultValue: PropTypes.any,
  placeholder: PropTypes.string,
  disabled: PropTypes.bool,
  children: PropTypes.node,
  onChange: PropTypes.func,
  type: PropTypes.string,
  name: PropTypes.string
}

export default Select
