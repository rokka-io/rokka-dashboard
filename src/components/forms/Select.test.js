import React from 'react'
import Select from '../../../src/components/forms/Select'
import renderer from 'react-test-renderer'

test('Select does render', () => {
  const component = renderer.create(<Select onChange={() => {}} />)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Select does render with options', () => {
  const component = renderer.create(
    <Select onChange={() => {}}>
      <option value="test">Test</option>
      <option value="test2">Test2</option>
    </Select>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Select does forward props', () => {
  const component = renderer.create(<Select foo="bar" onChange={() => {}} />)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Select takes only value if not disabled', () => {
  const component = renderer.create(<Select defaultValue="DefaultValue" />)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Select takes defaultValue if disabled', () => {
  const component = renderer.create(<Select defaultValue="DefaultValue" disabled />)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Select shows placeholder', () => {
  const component = renderer.create(<Select placeholder="Placeholder" onChange={() => {}} />)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Select shows placeholder with default value', () => {
  const component = renderer.create(
    <Select defaultValue="DefaultValue" placeholder="Placeholder" onChange={() => {}} />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Select takes name as placeholder if no placeholder given', () => {
  const component = renderer.create(
    <Select defaultValue="DefaultValue" name="Name" onChange={() => {}} />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Select takes type as placeholder if no placeholder & name given', () => {
  const component = renderer.create(
    <Select defaultValue="DefaultValue" type="Type" onChange={() => {}} />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Select is disabled if there is no onChange func', () => {
  const component = renderer.create(<Select />)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
