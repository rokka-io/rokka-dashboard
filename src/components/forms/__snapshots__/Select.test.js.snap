// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select does forward props 1`] = `
<select
  disabled={false}
  foo="bar"
  onChange={[Function]}
>
  <option />
</select>
`;

exports[`Select does render 1`] = `
<select
  disabled={false}
  onChange={[Function]}
>
  <option />
</select>
`;

exports[`Select does render with options 1`] = `
<select
  disabled={false}
  onChange={[Function]}
>
  <option />
  <option
    value="test"
  >
    Test
  </option>
  <option
    value="test2"
  >
    Test2
  </option>
</select>
`;

exports[`Select is disabled if there is no onChange func 1`] = `
<select
  disabled={true}
>
  <option />
</select>
`;

exports[`Select shows placeholder 1`] = `
<select
  disabled={false}
  onChange={[Function]}
  placeholder="Placeholder"
>
  <option>
    Placeholder
  </option>
</select>
`;

exports[`Select shows placeholder with default value 1`] = `
<select
  disabled={false}
  onChange={[Function]}
  placeholder="Placeholder (default: DefaultValue)"
>
  <option>
    Placeholder (default: DefaultValue)
  </option>
</select>
`;

exports[`Select takes defaultValue if disabled 1`] = `
<select
  disabled={true}
  placeholder=" (default: DefaultValue)"
  value="DefaultValue"
>
  <option>
     (default: DefaultValue)
  </option>
</select>
`;

exports[`Select takes name as placeholder if no placeholder given 1`] = `
<select
  disabled={false}
  name="Name"
  onChange={[Function]}
  placeholder="Name (default: DefaultValue)"
>
  <option>
    Name (default: DefaultValue)
  </option>
</select>
`;

exports[`Select takes only value if not disabled 1`] = `
<select
  disabled={true}
  placeholder=" (default: DefaultValue)"
  value="DefaultValue"
>
  <option>
     (default: DefaultValue)
  </option>
</select>
`;

exports[`Select takes type as placeholder if no placeholder & name given 1`] = `
<select
  disabled={false}
  onChange={[Function]}
  placeholder="Type (default: DefaultValue)"
  type="Type"
>
  <option>
    Type (default: DefaultValue)
  </option>
</select>
`;
